### PR TITLE
Testing out jupytext

### PR DESCRIPTION
**Pull request by elvijs**
* Adding Makefile commands for syncing notebooks as well as checking the sync
* Updating the requirements
* Drive-by: adding more gitignore dirs

Overall, I'm not convinced about jupytext so far. At first glance, it doesn't look like the sync checker is working correctly:
* I updated the script with additional bits that are not in the notebook, see https://github.com/ManchesterBioinference/BranchedGP/blob/elvijs/use-jupytext/notebooks/Hematopoiesis.py#L106 (note that this isn't reflected in the notebook, see https://github.com/ManchesterBioinference/BranchedGP/blob/elvijs/use-jupytext/notebooks/Hematopoiesis.ipynb
* However the checker doesn't pick up this difference. Instead it complains about some scrolling metadata, see below.

```
(.oldvenv) elvijs@elvijs-HP-Pavilion-Notebook:~/dev/BranchedGP$ make check_notebooks_synced 
jupytext --test-strict -x notebooks/*.ipynb --to py:percent
[jupytext] Reading notebooks/Hematopoiesis.ipynb in format ipynb
notebooks/Hematopoiesis.ipynb: Metadata differ on code cell #7: 
--- expected
+++ actual
@@ -1,3 +1 @@
-{
- "scrolled": true
-}
+{}
Cell content:
# Plot Monocle DDRTree space
genelist = ['FLT3','KLF1','MPO']
f, ax = plt.subplots(1, len(genelist), figsize=(10, 5), sharex=True, sharey=True)
for ig, g in enumerate(genelist):
        y = Y[g].values
        yt = np.log(1+y/y.max())
        yt = yt/yt.max()
        h = ax[ig].scatter(monocle['DDRTreeDim1'], monocle['DDRTreeDim2'],
                       c=yt, s=50, alpha=1.0, vmin=0, vmax=1)
        ax[ig].set_title(g)[jupytext] Reading notebooks/SamplingFromTheModel.ipynb in format ipynb
notebooks/SamplingFromTheModel.ipynb: Metadata differ on markdown cell #9: 
--- expected
+++ actual
@@ -1,3 +1 @@
-{
- "scrolled": true
-}
+{}
Cell content:
Sample the kernel[jupytext] Reading notebooks/SyntheticData.ipynb in format ipynb
notebooks/SyntheticData.ipynb: Metadata differ on code cell #4: 
--- expected
+++ actual
@@ -1,3 +1 @@
-{
- "collapsed": true
-}
+{}
Cell content:
datafile='syntheticdata/synthetic20.csv'
data = pd.read_csv(datafile, index_col=[0])
G = data.shape[1] - 2  # all data - time columns - state column
Y = data.iloc[:, 2:]
trueBranchingTimes = np.array([float(Y.columns[i][-3:]) for i in range(G)])make: *** [Makefile:20: check_notebooks_synced] Error 3
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -9,9 +9,17 @@ install:
...
 
+
 jupyter_server:
 	jupyter notebook $(NOTEBOOK_PATH)
 
+sync_notebooks:
+	jupytext --sync $(NOTEBOOK_PATH)/*.ipynb
+
+check_notebooks_synced:
+	jupytext --test-strict -x $(NOTEBOOK_PATH)/*.ipynb --to py:percent
```
Comment (alexisboukouvalas) [inline Makefile:20]: how about also doing

```
format_notebooks:
	jupytext --pipe 'isort - --treat-comment-as-code "# %%" --float-to-top' $(NOTEBOOK_PATH)/*.ipynb
	jupytext --sync --pipe black $(NOTEBOOK_PATH)/*.ipynb
```
This runs black and isort on the notebooks which is pretty cool

**Review by alexisboukouvalas — COMMENTED** [review]
Overall this is great. I am not sure how to fix the comparison issue.
I had an extensive look at https://jupytext.readthedocs.io/en/latest/using-cli.html and tried some options. For example this does not happen if we use the `md` format right?

There is also some documentation on metadata filters which may help ?

We can dig more into this, ignore the issue and assume it's all ok or not use jupytext.

File: Makefile (hunk 1 above)
Comment (elvijs) [reply Makefile:20]: Yeah, I like that! I'll do that as part of enabling black, isort and lint on the whole repo if that's OK.

**Comment by elvijs** [comment]
> I had an extensive look at https://jupytext.readthedocs.io/en/latest/using-cli.html and tried some options. For example this does not happen if we use the `md` format right?

I haven't tried the md format, but I'm a bit reluctant to use md as it's a bit less runnable (and blackable, lintable etc.) on its own. Let me give it a go and see what happens.

> There is also some documentation on metadata filters which may help ?

Thanks, I'll look into it.

> We can dig more into this, ignore the issue and assume it's all ok or not use jupytext.

Let me dig a bit more. It might be that I'm misreading the docs and am using the package incorrectly. 

Separately, even if the sync checks aren't perfect, the package could still be useful - the reviews are certainly easier.

**Comment by elvijs** [comment]
OK, I actually no longer think this is a big problem as Jupyter will alert you that the notebook is out of sync, see the attached screenshot.
Sync check in jupyter

Therefore, I think the risk of people looking at different versions of notebooks is sufficiently low and the changes seem worth it. I'm therefore promoting this to a full PR.

Summary of changes:
* Added Jupytext support (declared it in `test_requirements.txt`, updated the `README` and added useful `Makefile` commands).
* Paired up the existing notebooks.
* Drive-by: tidied up the `Makefile` a bit.

Note: you will probably find it easier to review this PR commit-by-commit.

File: Makefile [hunk 2]
```diff
@@ -1,28 +1,46 @@
...
 
 
-freeze_requirements:
-	pip freeze > $(TEST_REQUIREMENTS)
+#############################################
+# Commands for making sure code doesn't rot #
+#############################################
 
+test:
+	nosetests $(TEST_PATH)
```
Comment (alexisboukouvalas) [inline Makefile:43]: not really this PR but should we switch to pytest?

File: README.md [hunk 3]
```diff
@@ -62,10 +62,18 @@ Create a virtual environment, activate it and run `make install`.
 
 We use [Jupytext](https://github.com/mwouts/jupytext) to help version
 Jupyter notebooks.
-See in particular the documentation on [paired notebooks](https://jupytext.readthedocs.io/en/latest/#paired-notebooks).
+Each notebook corresponds to a Python script, which is easy to review.
+See also in the Jupytext documentation on
+[paired notebooks](https://jupytext.readthedocs.io/en/latest/#paired-notebooks).
+
```
Comment (alexisboukouvalas) [inline README.md:68]: How about we add a contribution section saying what is the process of submitting changes to code & notebooks?

For example when updating notebook, you need to rerun the whole notebook and then run `make pair_notebooks` ?

**Review by alexisboukouvalas — APPROVED** [review]
Look great!

File: Makefile (hunk 2 above)
Comment (elvijs) [reply Makefile:43]: I've certainly used pytest more than nosetests, so would feel more comfortable with it. What do you think @sumonahmedUoM ?

File: README.md (hunk 3 above)
Comment (elvijs) [reply README.md:68]: Yep, good suggestion, let me do that!

File: Makefile (hunk 2 above)
Comment (sumonahmedUoM) [reply Makefile:43]: Yes, we should switch to pytest, I am also comparatively more familiar with it. @elvijs @alexisboukouvalas

**Pull request merged**